beacon-chain/powchain/service.go

```diff
@@ -158,20 +207,136 @@ func (w *Web3Service) run(done <-chan struct{}) {
 		case VRClog := <-w.logChan:
-			// public key is the second topic from validatorRegistered log.
-			pubKeyLog := VRClog.Topics[1].Hex()
-			// Support user pubKeys with or without the leading 0x.
-			if pubKeyLog == w.pubKey || pubKeyLog[2:] == w.pubKey {
-				log.WithFields(logrus.Fields{
-					"publicKey": pubKeyLog,
-				}).Info("Validator registered in VRC with public key")
-				w.validatorRegistered = true
-				w.logChan = nil
-			}
+			w.processLog(VRClog)
+
 		}
 	}
 }
 
+func (w *Web3Service) processLog(VRClog gethTypes.Log) {
+	// Process logs according to their event signature.
+	if VRClog.Topics[0] == hashutil.Hash(depositEventSignature) {
+		w.processDepositLog(VRClog)
+		return
+	}
+
+	if VRClog.Topics[0] == hashutil.Hash(chainStartEventSignature) {
+		w.processChainStartLog(VRClog)
+		return
+	}
+
+	log.Debugf("Log is not of a valid event signature %#x", VRClog.Topics[0])
+}
+
+// processDepositLog processes the log which had been received from
+// the POW chain by trying to ascertain which participant deposited
+// in the contract.
+func (w *Web3Service) processDepositLog(VRClog gethTypes.Log) {
+	merkleRoot := VRClog.Topics[1]
+	depositData, MerkleTreeIndex, err := w.unPackDepositLogData(VRClog.Data)
+	if err != nil {
+		log.Errorf("Could not unpack log %v", err)
+		return
+	}
+	if err := w.saveInTrie(depositData, merkleRoot); err != nil {
+		log.Errorf("Could not save in trie %v", err)
+		return
+	}
+	depositInput, err := blocks.DecodeDepositInput(depositData)
+	if err != nil {
+		log.Errorf("Could not decode deposit input  %v", err)
+		return
+	}
+	index := binary.BigEndian.Uint64(MerkleTreeIndex)
+	log.WithFields(logrus.Fields{
+		"publicKey":         depositInput.Pubkey,
+		"merkle tree index": index,
+	}).Info("Validator registered in VRC with public key and index")
+}
+
+// processChainStartLog processes the log which had been received from
+// the POW chain by trying to determine when to start the beacon chain.
+func (w *Web3Service) processChainStartLog(VRClog gethTypes.Log) {
+	receiptRoot := VRClog.Topics[1]
+	timestampData, err := w.unPackChainStartLogData(VRClog.Data)
+	if err != nil {
+		log.Errorf("Unable to unpack ChainStart log data %v", err)
+		return
+	}
+	if w.depositTrie.Root() != receiptRoot {
+		log.Errorf("Receipt root from log doesn't match the root saved in memory %#x", receiptRoot)
+		return
+	}
+
+	timestamp := binary.BigEndian.Uint64(timestampData)
+	if uint64(time.Now().Unix()) < timestamp {
+		log.Errorf("Invalid timestamp from log expected %d > %d", time.Now().Unix(), timestamp)
+	}
+
+	chainStartTime := time.Unix(int64(timestamp), 0)
+	log.WithFields(logrus.Fields{
+		"ChainStartTime": chainStartTime,
+	}).Info("Minimum Number of Validators Reached for beacon-chain to start")
+}
+
+// unPackDepositLogData unpacks the data from a log using the abi decoder.
+func (w *Web3Service) unPackDepositLogData(data []byte) (depositData []byte, merkleTreeIndex []byte, err error) {
```

Comment:
I don't think this and the one below should be methods of the web3 service but instead just standalone functions that take in the abi as an argument

Reply:
+1

Reply:
unpack is one word, please change to `unpackDepositLogData`

Reply:
alright I moved it into the utils package